### ISSUE TITLE
Incorrect type promotion in co19_2/Language/Expressions/Logical_Boolean_Expressions/type_t16.dart

### ISSUE THREAD

**Issue by alexmarkov** [dart-lang/sdk#43640]
`co19_2/Language/Expressions/Logical_Boolean_Expressions/type_t16.dart` test has the following code:

```dart
f(var x) {
  x is D && skyIsBlue(x = new C()) ? x.f() : null;
}
```

Kernel for this function:
```
  static method f(dynamic x) → dynamic {
    x is typ::D* && typ::skyIsBlue(x = new typ::C::•()) as{TypeError,ForDynamic} core::bool* ?{dynamic} x{typ::D*}.{typ::D::f}() : null;
  }
```

In this function variable `x` is incorrectly promoted to type `D*`, while it was actually modified.
Dynamic call `x.f()` is incorrectly converted to an interface call with incompatible interface target.
This results in an incorrect code in AOT (or crash after https://dart-review.googlesource.com/c/sdk/+/165625).

@johnniwinther @stefantsov @stereotype441

**Comment by stereotype441** [comment]
I will investigate ASAP

**Comment by stereotype441** [comment]
What's going on here is a bug in the front end's implementation of legacy type promotion (which I implemented way back in April of 2017!)  It turns out that the front end doesn't implement the proper rules for when assignments defeat type promotion.  The rules it's supposed to implement are:
- When an the condition of an `if` or `?:` attempts to promote a variable to a type, but the body of the `then` contains an assignment to that variable, the promotion doesn't happen.
- When the LHS of an `&&` promotes a variable to a type, but the RHS contains an assignment to that variable, the promotion doesn't happen.

The rule it actually implements is:
- When there is an assignment to a variable, promotion of that variable is disabled in the innermost enclosing "type promotion scope", where a type promotion scope is the RHS of an `&&` or `||`, or the part of an `if` or `?:` that follows the condition.

Amazingly, as different as these rules are, this bug apparently went undetected for 3 years!  Honestly, I'm kind of baffled.  I guess it's mostly because the front end defeats fewer promotions than it should (which means that by and large, the failure mode is that code is incorrectly accepted rather than incorrectly rejected), and so fortunately people who code using the analyzer never write the incorrect code in the first place.

So, in the example from the bug report, there are two type promotion scopes:
```dart
f(var x) {
  x is D && skyIsBlue(x = new C()) ? x.f() : null;
  //        ^------scope #1------^   ^-scope #2-^
}
```
The assignment defeats promotion in scope #1 but not in scope #2, so `x` is incorrectly considered promoted at the site of `x.f()`.

I'll start working on a fix.

**Comment by stereotype441** [comment]
FYI, fixing this bug is technically a breaking change (though hopefully the impact will be small due to the fact that the analyzer implements promotion correctly).  I'll work with @leafpetersen to figure out whether we need to go through the breaking change process on this.

**Comment by leafpetersen** [comment]
Technically I believe that this fix could change the behavior of working code.  It's a very unlikely case: you need code that is valid with and without the promotion (rare already), and the difference in the type needs to either affect something statically (e.g. extension method resolution) or dynamically (by being reified).  I would be amazed if any code actually hits that trifecta, but it is technically possible.

@stereotype441 I would suggest that you land the fix with a note in the CHANGELOG about the breaking change.  

cc @davidmorgan as a heads up on any possible internal breakage (very unlikely).

We should file a breaking change request to give people time to raise any concerns before the next stable release - I'm happy to do that.

**Comment by stereotype441** [comment]
Fixed by 2c40d6a22a4a63eca2cf4515297da35405d95dbb
